### PR TITLE
repo: refactored repository interfaces

### PR DESCRIPTION
**Pull request by jkowalski**
- `repo.Repository` is now read-only and only has methods that can be supported over kopia server
- `repo.RepositoryWriter` has read-write methods that can be supported over kopia server
- `repo.DirectRepository` is read-only and contains all methods of `repo.Repository` plus some low-level methods for data inspection
- `repo.DirectRepositoryWriter` contains write methods for `repo.DirectRepository`

- `repo.Reader` removed and merged with `repo.Repository`
- `repo.Writer` became `repo.RepositoryWriter`
- `*repo.DirectRepository` struct became `repo.DirectRepository`
  interface

Getting `{Direct}RepositoryWriter` requires using `NewWriter()` or `NewDirectWriter()` on a read-only repository and multiple simultaneous writers are supported at the same time, each writing to their own indexes and pack blobs.

`repo.Open` returns `repo.Repository` (which is also `repo.RepositoryWriter` for legacy purposes)

### CODE REVIEW

File: repo/splitter/splitter.go [hunk 1]
```diff
@@ -7,6 +7,11 @@ import (
 
 const (
 	splitterSlidingWindowSize = 64
+
+	splitterSize1MB = 1 << 20
+	splitterSize2MB = 2 << 20
+	splitterSize4MB = 4 << 20
+	splitterSize8MB = 8 << 20
```
Comment (julio-lopez) [inline repo/splitter/splitter.go:14]: nit: given that this is in the splitter package and not exported, they could be simply named `sizeXMB`

File: repo/content/content_manager.go [hunk 2]
```diff
@@ -214,21 +214,6 @@ func (bm *WriteManager) addToPackUnlocked(ctx context.Context, contentID ID, dat
...
-	fp := append([]*pendingPackInfo(nil), bm.failedPacks...)
-	for _, pp := range fp {
-		formatLog(ctx).Debugf("retry-write %v", pp.packBlobID)
-
-		if err := bm.writePackAndAddToIndex(ctx, pp, true); err != nil {
-			bm.unlock()
-			return errors.Wrap(err, "error writing previously failed pack")
-		}
-	}
-
```
[the remark this answers is not in the document]
Comment (jkowalski) [reply repo/content/content_manager.go:231]: not intentional, will fix

File: internal/server/api_manifest.go [hunk 3]
```diff
@@ -41,9 +42,14 @@ func (s *Server) handleManifestGet(ctx context.Context, r *http.Request, body []
 }
 
 func (s *Server) handleManifestDelete(ctx context.Context, r *http.Request, body []byte) (interface{}, *apiError) {
+	rw, ok := s.rep.(repo.RepositoryWriter)
+	if !ok {
+		return nil, notFoundError("manifest not found")
```
[the remark this answers is not in the document]
Comment (jkowalski) [reply internal/server/api_manifest.go:47]: fixed in all places.

File: repo/splitter/splitter.go (hunk 1 above)
Comment (jkowalski) [reply repo/splitter/splitter.go:14]: I already pushed that commit independently as I was testing linter behavior.

File: repo/repository.go [hunk 4]
```diff
@@ -227,14 +313,66 @@ func (r *DirectRepository) RefreshPeriodically(ctx context.Context, interval tim
...
+
+	resultErr := cb(w)
+
+	if resultErr == nil || opt.FlushOnFailure {
+		if err := w.Flush(ctx); err != nil {
+			return errors.Wrap(err, "error flushing writer")
+		}
+	}
+
+	if err := w.Close(ctx); err != nil {
```
[the remark this answers is not in the document]
Comment (jkowalski) [reply repo/repository.go]: fixed

File: repo/content/committed_read_manager.go [hunk 5]
```diff
@@ -319,6 +322,32 @@ func (sm *SharedManager) setupReadManagerCaches(ctx context.Context, caching *Ca
 	return nil
 }
 
+// AddRef adds a reference to shared manager to prevents its closing on Release().
+func (sm *SharedManager) addRef() {
+	atomic.AddInt32(&sm.refCount, 1)
```
[the remark this answers is not in the document]
Comment (jkowalski) [reply repo/content/committed_read_manager.go:332]: not really, but it's a coding bug:

```
r = repo.Open()
r.Close()
r.NewWriter()
```

added nice panic

File: repo/content/committed_read_manager.go [hunk 6]
```diff
@@ -335,6 +340,8 @@ func (sm *SharedManager) release(ctx context.Context) error {
 		return nil
 	}
 
+	sm.closed = true
```
Comment (julio-lopez) [inline repo/content/committed_read_manager.go:343]: There is no guarantee that this write becomes visible elsewhere.

**Comment by julio-lopez** [comment]
Fixes look good.

See comment about unprotected write to `sm.closed`

File: repo/content/committed_read_manager.go (hunk 6 above)
Comment (jkowalski) [reply repo/content/committed_read_manager.go]: fixed

**Review by julio-lopez — APPROVED** [review]

**Pull request merged**